### ISSUE TITLE
[peer-mgr] Allocate more peers to downloading torrents

### ISSUE THREAD

**Issue by tearfur** [transmission/transmission#6515]
## tl:dr

Transmission should prioritise peer allocation for downloading torrents, disconnecting peers in seeding torrents to make room when necessary.

## Details

In the nightlies, with default peer limit settings and a lot of seeding torrents, the downloading torrents usually won't reach the per-torrent peer limit. However, as soon as you pause the seeding torrents, the downloading torrents fill up the per-torrent peer limit.

This is because since 8169d524eab8aeb5cf16be12182e83dde109ffd4, the seeding torrents are usually always connected to some peers, even if there are no active transfers. These peers occupy the global peer limit, and there are currently no mechanisms in place to disconnect them to make space for torrents that need more peers.

So, even if other torrents have per-torrent peer quota to spare, they still can't connect to more peers because of the global peer limit.

I think 8169d524eab8aeb5cf16be12182e83dde109ffd4 by itself is a good change, so the thing to do here is not to revert that, but to implement a way for torrents to "steal" peer quota from others.

### Which application of Transmission?

None

### Which version of Transmission?

239478925f3a6a20dda7f0c3f4a317e8ae15a7f5

**Comment by Pentaphon** [comment]
> default peer limit settings

We should also rethink the current default global peer limit of 200. Maybe 300 at the very least?

As for the mechanism that could be used... Perhaps set a prioritization so that older non-seeding connections can be used to download? No idea how long these connections typically last but instead of making them all the same priority status, you can create a 2-tier priority status so some connections are always kept but others can be disconnected to download after a certain time?

**Comment by Coeur** [comment]
I apparently set mine to 9999 peers globally and 999 peers per torrent. So yes, I think at some point I was hitting the limit, and I workaround it like that. I'm not sure what could go wrong with 9999 peers.

**Comment by Coeur** [comment]
>disconnecting peers in seeding torrents

Well, I would keep the logic as:
"**while (too many peers) { disconnect 1 peer from the torrent with the highest number of connected_peers }**".

Although, I do not know how to balance that with **tr_priority_t** (which is -1, 0 or +1) which is presently defined on the bandwidth of each torrent?

Maybe:
"**while (too many peers) { disconnect 1 peer from the torrent with the highest number of (connected_peers * (1.0 - 0.1*get_priority))  }**".
(aka: if priority is high, we rank for disconnection as connected_peers * 0.9, and if priority is low, we rank for disconnection as connected_peers * 1.1)

**Comment by Coeur** [comment]
Actually, there may be a risk of too frequent peers disconnection-reconnection when doing this kind of load balancing.
So another idea, to avoid having to disconnect peers all the time, is that:
- the first 90% of the total allowed number of peers is attributed as usual
- the last 10% of the total allowed number of peers is only given to torrents with fewer peers than others

**Comment by Pentaphon** [comment]
> Actually, there may be a risk of too frequent peers disconnection-reconnection when doing this kind of load balancing.
> So another idea, to avoid having to disconnect peers all the time, is that:

You may have to experiment and do some trial and error with that. 90:10 might be too high. Could be better at 80:20 or 70:30. I do like this idea of splitting peers into two tiers.

I still think we need to re-consider our default global connection of 200 as well. Many people don't like to change their settings and having more sane defaults benefits everybody.  300 would be a good start.

**Comment by tearfur** [comment]
> Actually, there may be a risk of too frequent peers disconnection-reconnection when doing this kind of load balancing.

I think we can get around this by only allowing downloading torrents to steal peers from seeding torrents. In other words, no one can steal peers from downloading torrents, and seeding torrents cannot steal peers from anyone.

**Comment by reardonia** [comment]
Having played around today in the get_peer_candidates() I'm wondering what you're thinking now?  Current code is a wee bit broken and doesn't properly favor new downloading torrents but should.

**Comment by Pentaphon** [comment]
> Current code is a wee bit broken and doesn't properly favor new downloading torrents but should.

At the very least it should meet what is proper because the complaint I typically see about Transmission is that it is not aggressive enough when downloading torrents. I think the current code is a holdover from when hardware wasn't very fast but these days, Transmission is running on much better home hardware and should favor downloading more.
